### PR TITLE
Add configurable Mixer transport error retry

### PR DESCRIPTION
**Pull request by howardjohn**
Adds annotations for the number of retries, base wait time, and max wait
time to configure Mixer transport error retry policy. If values are not
provided, they will be left unset; defaults will be provided in
istio/proxy.

### CODE REVIEW

File: pilot/pkg/networking/plugin/mixer/mixer.go [hunk 1]
```diff
@@ -53,6 +54,15 @@ const (
 	// workload-bound annotations for controlling mixerfilter
 	policyCheckAnnotation = "policy.istio.io/check"
 
+	// workload-bound annotation for max retries on transport error
```
Comment (costinm) [inline pilot/pkg/networking/plugin/mixer/mixer.go]: pilot/pkg/model/context.go - in general node metadata ( like annotations ) should go there. Can move it in separate PR, I see others got in the wrong place.

File: pilot/pkg/networking/plugin/mixer/mixer.go [hunk 2]
```diff
@@ -53,6 +54,15 @@ const (
...
+	policyCheckRetriesAnnotation = "policy.istio.io/checkRetries"
+
+	// workload-bound annotation for base time to wait between retries
```
Comment (costinm) [inline pilot/pkg/networking/plugin/mixer/mixer.go]: Not sure a user would understand from comment or name. Maybe few more words, example value, etc ?

Comment (costinm) [reply pilot/pkg/networking/plugin/mixer/mixer.go]: And format - duration is not what most users expects.

File: pilot/pkg/networking/plugin/mixer/mixer.go [hunk 3]
```diff
@@ -247,25 +257,54 @@ func buildUpstreamName(address string) string {
...
+			policy = mccpb.FAIL_OPEN
+		}
+	}
+
+	networkFailPolicy := &mccpb.NetworkFailPolicy{Policy: policy}
+
+	if annotation, ok := node.Metadata[policyCheckRetriesAnnotation]; ok {
+		retries, err := strconv.Atoi(annotation)
+		if err != nil {
+			log.Warnf("unable to parse retry limit %v.", annotation)
```
Comment (kyessenov) [inline pilot/pkg/networking/plugin/mixer/mixer.go]: cc @costinm : is there a log dampener in pilot? this might be logging on every request

**Comment by mandarjog** [comment]
What are the defaults provided in proxy? Perhaps pilot should provide it’s own default when num retries is set. It is too much to require to set all 3 annotations.

File: pilot/pkg/networking/plugin/mixer/mixer_test.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+	"github.com/gogo/protobuf/types"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	mccpb "istio.io/api/mixer/v1/config/client"
+	"istio.io/istio/pilot/pkg/model"
+	context "istio.io/istio/pilot/pkg/model"
+)
+
+func TestTransportConfig(t *testing.T) {
```
Comment (costinm) [inline pilot/pkg/networking/plugin/mixer/mixer_test.go]: I understand wanting to move fast and have good code coverage - but for this kind of change an e2e test would be far more appropriate. Is the annotation really propagated from pod annotation all the way to the mixer, and the right config generated ? 

I would settle for just an example config, with the annotations set, in the tools/ repo
where we do the real testing - so at least it is easier to manually test this until 
it is automated, and we can get some meaningful metrics. 

Maybe - this test only checks that if you set a value in a hashmap - you get it back.

File: pilot/pkg/networking/plugin/mixer/mixer.go (hunk 3 above)
Comment (costinm) [reply pilot/pkg/networking/plugin/mixer/mixer.go]: Right - and nobody is really watching the logs.

How about a prometheus variable - tracking invalid annotations of each kind for example - which can be at least acted on.

File: pilot/pkg/networking/plugin/mixer/mixer.go [hunk 5]
```diff
@@ -247,25 +257,54 @@ func buildUpstreamName(address string) string {
...
+		} else {
+			networkFailPolicy.MaxRetry = uint32(retries)
+		}
+	}
+
+	if annotation, ok := node.Metadata[policyCheckBaseRetryWaitTimeAnnotation]; ok {
+		dur, err := time.ParseDuration(annotation)
+		if err != nil {
+			log.Warnf("unable to parse base retry wait time %v.", annotation)
```
Comment (kyessenov) [inline pilot/pkg/networking/plugin/mixer/mixer.go]: nit: use `%q` to print string variables in logs in quotes

**Comment by costinm** [comment]
Actually: please fix at least Mandar's comment, maybe set some reasonable defaults if the first annotation is present ? 3 annotations seem too much, we already have too many knobs.

**Comment by howardjohn** [comment]
> What are the defaults provided in proxy? Perhaps pilot should provide it’s own default when num retries is set. It is too much to require to set all 3 annotations.

@mandarjog The defaults are 80ms and 1000ms. https://github.com/istio/proxy/pull/2113. If they don't provide anything then we don't set anything here so proxy will set the defaults. This applies even if we provide the number of retries, but not the times, so they don't need to set all 3. I was avoiding setting defaults here so we don't have different defaults in different places for the same thing, but I can change it if you want

**Comment by kyessenov** [comment]
Logic seems reasonable. I'd actually prefer reusing envoy's config policy for this purpose. The difference is that there is only one upper bound and one max retry knob. The retries never exceeds the timeout. I think it's equivalent to setting both initial and max to same value.

cc @duderino : I think we do need a test that validates the generated config against an actual proxy. Something in the style of `mixer/test/client` framework with hard-coded config.

File: pilot/pkg/networking/plugin/mixer/mixer_test.go (hunk 4 above)
Comment (kyessenov) [reply pilot/pkg/networking/plugin/mixer/mixer_test.go]: This is similar to pod policy annotation. I manually verified that propagation works. I think @howardjohn also needs to manually verify this e2e. It's not entirely obvious how to simulate and validate retries. Perhaps, killing policy and monitoring the proxy counters?

**Comment by kyessenov** [comment]
/lgtm
/approve

Logic looks right. Clean-ups can follow.

**Comment by mandarjog** [comment]
In a followup PR, please add default values to pilot and document them in the annotations.
/lgtm

**Comment by duderino** [comment]
I failure in ci/circileci: e2e-pilot-auth-v1alpha3-v2.  Rerunning cicleci workflow to see if it's a flake:

TestIngressGateway503DuringRuleChange - pilot Failed

```
gateway "istio-ingressgateway" created
destinationrule "destination-rule-c" created
virtualservice "gateway-bound-route" created
destinationrule "destination-rule-c" configured
virtualservice "gateway-bound-route" configured
destinationrule "destination-rule-c" configured
virtualservice "gateway-bound-route" deleted
destinationrule "destination-rule-c" deleted
gateway "istio-ingressgateway" deleted
ingressgateway_test.go:258: Got non 200 status code while changing rules: map[200:499 503:1]
```

**Comment by duderino** [comment]
@costinm @howardjohn @mandarjog @kyessenov 

Unless anyone thinks the failures are real problems, I will merge this at 12pm PT Tuesday 2/19 (today)

**Comment by kyessenov** [comment]
From my end, it's good for 1.1. I think we need to move config into another placeholder in 1.2, since this change makes config hard to cache. It has conflicts with #11469, for example (cc @rshriram).

The more uniform we can make the config across proxies, the less work it is for pilot to generate it. Since mixerfilter config is spread all over place in LDS/RDS, that is a lot of work and it would be better to put retry config into cluster metadata using just CDS.
